README.md
A beautiful calendar library for cross platforms.

## Maxx Calendar

This package provides two month view modes.

![Group 4 (1)](https://user-images.githubusercontent.com/64951609/247869785-47fa9aa8-c91a-420a-bebc-b00ae5082cc8.png)



## Features

- Switch MonthView modes between events with visible texts and events with dots.
- In built buttons for switching months
- Null safe
- Border colors, text colors and events background colors can be customized easily.
- In built buttons for month switching can be hidden.
- Callbacks for month changes and event clicks.

## Getting started

Just add the widget to your widget tree and go ahead !

## Usage

```
//create Events (read the comments carefully)
List<List<MaxxEventModel>> events = [
      [
        MaxxEventModel(
            id: "unique id here if you want or the date format you want",
            title: "event 1 name",
            bgColor: Colors.cyan,
            textColor: Colors.white),
        MaxxEventModel(
            id: "unique id here if you want or the date format you want",
            title: "event 2 name",
            bgColor: Colors.cyan,
            textColor: Colors.white)
      ], //Adds this events to day 1 of the current month
      <MaxxEventModel>[].toList() //this adds empty events for day 2
      //You must add empty events for days which should not have anything
    ];

Expanded(
         child: MaxxCalendar(calType: CalType.MONTHVIEW,
                tileBorderColor: Colors.indigoAccent,
                tileDateColor: Colors.black,
                events: events, //fetch new events from your remote repo on date change
                onDateClick: (date){},
                onDateChangeListener: (date){},
                currMonth: 6, //sets the initial month
                currYear: 2023, //sets the initial year
                hideMonthNavigationButtons: false),
         )
```





CHANGELOG.md
## 0.0.1

* Initial Release
* Add more views

## 0.0.2

* bug fixes

## 1.0.0

* Added usage

## 1.0.1 - 1.0.11

* Code Optimizations

test/maxx_calendar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:maxx_calendar/maxx_calendar.dart';


lib/widget/views.dart

import 'package:flutter/material.dart';
import 'package:auto_size_text/auto_size_text.dart';
import 'package:google_fonts/google_fonts.dart';




Widget headerText(String text, double size, {Color color = Colors.white}) {
  return AutoSizeText(
    text,
    style: GoogleFonts.poppins(
        textStyle: TextStyle(
            color: color, fontWeight: FontWeight.bold, fontSize: size)),
    minFontSize: size,
  );
}

Widget titleText(String text, double size, {Color textColor = Colors.black, TextAlign textAlign = TextAlign.start}) {
  return AutoSizeText(
    text,
    textAlign: textAlign,
    style: GoogleFonts.poppins(
        textStyle: TextStyle(color: textColor
            , fontSize: size)),
    minFontSize: 10,
    overflow: TextOverflow.ellipsis,
    maxFontSize: size,
  );
}

Widget paragraphText(String text, {double size = 8, Color textColor = Colors.black, TextAlign alignment = TextAlign.start}) {
  return AutoSizeText(
    text,
    style: GoogleFonts.poppins(
        textStyle: TextStyle(color: textColor, fontSize: size,)),
    textAlign: alignment,
    minFontSize: size,
    overflow: TextOverflow.ellipsis,
    softWrap: true,
  );
}


lib/widget/calendar.dart
import 'package:flutter/material.dart';
import 'package:flutter_animate/flutter_animate.dart';

import 'package:intl/intl.dart';
import 'package:maxx_calendar/widget/cal_model.dart';
import 'package:maxx_calendar/widget/calendar_tile.dart';
import 'package:maxx_calendar/widget/views.dart';

import 'calendar_dot_tile.dart';
import 'caltype.dart';
import 'constants.dart';

class MaxxCalendar extends StatefulWidget {
  DateTime now = DateTime.now();
  DateFormat formatter = DateFormat('EEE');
  DateFormat defaultFormat = DateFormat("dd-MM-yyyy");
  var calType = CalType.MONTHVIEW;
  Color tileBorderColor = Colors.black54;
  Color tileDateColor = Colors.black54;
  var currMonth = DateTime.now().month;
  var currYear = DateTime.now().year;
  var firstDayOfMonth = "mon";
  var firstDaySpacing = 0;
  var endDaySpacing = 0;
  var emptyEvents = <List<MaxxEventModel>>[];
  var hideMonthNavigationButtons = false;

  var events = <List<MaxxEventModel>>[];

  var resultEvents = <List<MaxxEventModel>>[];
  Function(Map<String, List<MaxxEventModel>>) onDateClick;
  Function(String) onDateChangeListener;

  MaxxCalendar(
      {Key? key,
      required this.calType,
      required this.tileBorderColor,
      required this.tileDateColor,
      required this.events,
      required this.onDateClick,
      required this.onDateChangeListener,
      required this.currMonth,
      required this.currYear,
      required this.hideMonthNavigationButtons})
      : super(key: key);

  @override
  State<MaxxCalendar> createState() => _MaxxCalendarState();
}

class _MaxxCalendarState extends State<MaxxCalendar> {
  static int getDaysInMonth(int year, int month) {
    if (month == DateTime.february) {
      final bool isLeapYear =
          (year % 4 == 0) && (year % 100 != 0) || (year % 400 == 0);
      return isLeapYear ? 29 : 28;
    }
    const List<int> daysInMonth = <int>[
      31,
      -1,
      31,
      30,
      31,
      30,
      31,
      31,
      30,
      31,
      30,
      31
    ];
    return daysInMonth[month - 1];
  }

  void createFinalList() {
    widget.firstDayOfMonth =
        widget.formatter.format(DateTime(widget.currYear, widget.currMonth, 0));

    widget.firstDaySpacing = Constants.firstDayOffset(
        widget.currYear,
        (widget.currMonth),
        const DefaultMaterialLocalizations());

    if (widget.firstDaySpacing > 0) {
      for (var i = 1; i <= widget.firstDaySpacing; i++) {
        widget.emptyEvents.add([]);
      }
    }

    widget.resultEvents.addAll(widget.emptyEvents);

    if (widget.events.length <=
        getDaysInMonth(widget.currYear, widget.currMonth)) {
      widget.resultEvents.addAll(widget.events);
    } else {
      throw ("Events size : ${widget.events.length} - Events are more than the number of days present in current month");
    }

    var tempEndspacing = (getDaysInMonth(widget.currYear, widget.currMonth) +
        widget.firstDaySpacing);
    widget.endDaySpacing = 0;
    if ((tempEndspacing) % 7 != 0) {
      for (var i = 0; i < 10; i++) {
        widget.endDaySpacing++;
        if ((tempEndspacing + widget.endDaySpacing) % 7 == 0) {
          break;
        }
      }
    }
  }

  @override
  void initState() {
    super.initState();
    createFinalList();
  }

  @override
  Widget build(BuildContext context) {

    void onNextBtnClick() {
      setState(() {
        widget.resultEvents = [];
        widget.emptyEvents = [];
        widget.currMonth != 12
            ? {widget.currMonth = (widget.currMonth + 1)}
            : {widget.currMonth = 1, widget.currYear = widget.currYear + 1};

        createFinalList();
      });
      widget.onDateChangeListener("${widget.currMonth}-${widget.currYear}");
    }

    void onPrevBtnClick() {
      setState(() {
        widget.resultEvents = [];
        widget.emptyEvents = [];
        widget.currMonth != 1
            ? {(widget.currMonth = widget.currMonth - 1)}
            : {widget.currMonth = 12, widget.currYear = widget.currYear - 1};
        createFinalList();
      });
      widget.onDateChangeListener("${widget.currMonth}-${widget.currYear}");
    }

    void invokeOnclick(Map<String, List<MaxxEventModel>> events) {
      widget.onDateClick(events);
    }

    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 5, vertical: 5),
          child: widget.hideMonthNavigationButtons == true
              ? const SizedBox()
              : Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    InkWell(
                      radius: 15,
                      borderRadius: BorderRadius.circular(15),
                      onTap: () {
                        onPrevBtnClick();
                      },
                      child: const Padding(
                        padding: EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.arrow_back_ios_new,
                          size: 15,
                        ),
                      ),
                    ),
                    paragraphText(
                        "${Constants.getCalendarMonth(widget.currMonth).toString()} ${widget.currYear}",
                        size: 12),
                    InkWell(
                      borderRadius: BorderRadius.circular(15),
                      child: const Padding(
                        padding: EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.arrow_forward_ios,
                          size: 15,
                        ),
                      ),
                      onTap: () {
                        onNextBtnClick();
                      },
                    ),
                  ],
                ),
        ),
        SizedBox(
          height: 16,
          child: GridView.count(
            crossAxisCount: 7,
            children: [
              paragraphText("Mon", alignment: TextAlign.center),
              paragraphText("Tue", alignment: TextAlign.center),
              paragraphText("Wed", alignment: TextAlign.center),
              paragraphText("Thu", alignment: TextAlign.center),
              paragraphText("Fri", alignment: TextAlign.center),
              paragraphText("Sat", alignment: TextAlign.center),
              paragraphText("Sun", alignment: TextAlign.center),
            ],
          ),
        ),
        Expanded(
          flex: 9,
          child: Animate(
            effects: const [FadeEffect()],
            child: GridView.builder(
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    childAspectRatio:
                        widget.calType == CalType.MONTHVIEW ? 8 / 16 : 1 / 1,
                    crossAxisSpacing: 0,
                    mainAxisSpacing: 0,
                    mainAxisExtent:
                        widget.calType == CalType.MONTHVIEW ? 84 : 64,
                    crossAxisCount: 7),
                itemCount: getDaysInMonth(widget.currYear, widget.currMonth) +
                    widget.firstDaySpacing +
                    widget.endDaySpacing,
                itemBuilder: (BuildContext ctx, index) {
                  var item = ((widget.resultEvents.length) - 1) >= index
                      ? widget.resultEvents[index]
                      : <MaxxEventModel>[];
                  return Animate(
                    effects: [FadeEffect(delay: 500.ms)],
                    child: widget.calType == CalType.MONTHVIEW
                        ? CalendarTile(
                            title: ((index - widget.firstDaySpacing < 0
                                    ? "${getDaysInMonth(widget.currMonth == 1 ? (widget.currYear - 1) : (widget.currYear), (widget.currMonth - 1 < 1 ? 12 : widget.currMonth - 1)) + 1 - (widget.firstDaySpacing) + index}"
                                    : index >=
                                            (getDaysInMonth(widget.currYear,
                                                    widget.currMonth) +
                                                widget.firstDaySpacing)
                                        ? "${index + 1 - (getDaysInMonth(widget.currYear, widget.currMonth) + widget.firstDaySpacing)}"
                                        : "${(index + 1) - widget.firstDaySpacing}"))
                                .toString(),
                            events: item,
                            onClick: (events) {
                              invokeOnclick(events);
                            },
                            isEnabled: (index - widget.firstDaySpacing < 0
                                ? false
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? false
                                    : true),
                            date: (index - widget.firstDaySpacing < 0
                                ? "" //previous month dates
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? "" //next month dates
                                    : "${index - widget.firstDaySpacing + 1 < 10 ? "0${index - widget.firstDaySpacing + 1}" : index - widget.firstDaySpacing + 1}-${widget.currMonth < 10 ? "0${widget.currMonth}" : widget.currMonth}-${widget.currYear}"),
                            todaysDate:
                                widget.defaultFormat.format(DateTime.now()),
                            tileBorderColor: widget.tileBorderColor,
                            tileTextColor: widget.tileDateColor,
                          )
                        : CalendarDotTile(
                            title: ((index - widget.firstDaySpacing < 0
                                    ? "${getDaysInMonth(widget.currMonth == 1 ? (widget.currYear - 1) : (widget.currYear), (widget.currMonth - 1 < 1 ? 12 : widget.currMonth - 1)) + 1 - (widget.firstDaySpacing) + index}"
                                    : index >=
                                            (getDaysInMonth(widget.currYear,
                                                    widget.currMonth) +
                                                widget.firstDaySpacing)
                                        ? "${index + 1 - (getDaysInMonth(widget.currYear, widget.currMonth) + widget.firstDaySpacing)}"
                                        : "${(index + 1) - widget.firstDaySpacing}"))
                                .toString(),
                            events: item,
                            onClick: (events) {
                              invokeOnclick(events);
                            },
                            isEnabled: (index - widget.firstDaySpacing < 0
                                ? false
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? false
                                    : true),
                            date: (index - widget.firstDaySpacing < 0
                                ? "" //previous month dates
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? "" //next month dates
                                    : "${index - widget.firstDaySpacing + 1 < 10 ? "0${index - widget.firstDaySpacing + 1}" : index - widget.firstDaySpacing + 1}-${widget.currMonth < 10 ? "0${widget.currMonth}" : widget.currMonth}-${widget.currYear}"),
                            todaysDate:
                                widget.defaultFormat.format(DateTime.now()),
                            tileBorderColor: widget.tileBorderColor,
                            tileTextColor: widget.tileDateColor,
                          ),
                  );
                }),
          ),
        ),
      ],
    );
  }
}


lib/widget/constants.dart
import 'package:flutter/material.dart';

class Constants {
  static String? getCalendarMonth(int month) {

    switch (month) {
      case 1:
        {
          return "January";
        }
      case 2:
        {
          return "February";
        }
      case 3:
        {
          return "March";
        }
      case 4:
        {
          return "April";
        }
      case 5:
        {
          return "May";
        }
      case 6:
        {
          return "June";
        }
      case 7:
        {
          return "July";
        }
      case 8:
        {
          return "August";
        }
      case 9:
        {
          return "September";
        }
      case 10:
        {
          return "October";
        }
      case 11:
        {
          return "November";
        }
      case 12:
        {
          return "December";
        }
    }
    return null;
  }



  static int firstDayOffset(int year, int month, MaterialLocalizations localizations) {
    // 0-based day of week for the month and year, with 0 representing Monday.
    final int weekdayFromMonday = DateTime(year, month).weekday - 2 ;

    // 0-based start of week depending on the locale, with 0 representing Sunday.
    int firstDayOfWeekIndex = localizations.firstDayOfWeekIndex;

    // firstDayOfWeekIndex recomputed to be Monday-based, in order to compare with
    // weekdayFromMonday.
    firstDayOfWeekIndex = (firstDayOfWeekIndex - 1) % 7;

    // Number of days between the first day of week appearing on the calendar,
    // and the day corresponding to the first of the month.
    return (weekdayFromMonday - firstDayOfWeekIndex) % 7;
  }

  static int? getSpacingForFirstDayOfMonth(String day) {


    switch (day.toLowerCase()) {
      case "mon":
        {
          return 0;
        }
      case "tue":
        {
          return 1;
        }
      case "wed":
        {
          return 2;
        }
      case "thu":
        {
          return 3;
        }
      case "fri":
        {
          return 4;
        }
      case "sat":
        {
          return 5;
        }
      case "sun":
        {
          return 6;
        }
    }
    return null;
  }
}


lib/widget/calendar_tile.dart
import 'package:flutter/material.dart';
import 'package:maxx_calendar/widget/cal_model.dart';
import 'package:maxx_calendar/widget/views.dart';

//ignore: must_be_immutable
class CalendarTile extends StatefulWidget {
  String title;
  List<MaxxEventModel> events = [];
  Function(Map<String, List<MaxxEventModel>>) onClick;
  bool isEnabled;
  String date;
  String todaysDate;
  Color tileTextColor;
  Color tileBorderColor;

  CalendarTile(
      {Key? key, required this.title,
      required this.events,
      required this.onClick,
      required this.isEnabled,
      required this.date,
      required this.todaysDate,
      required this.tileTextColor,
      required this.tileBorderColor}) : super(key: key);

  @override
  State<CalendarTile> createState() => _CalendarTileState();
}

class _CalendarTileState extends State<CalendarTile> {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
      child: InkWell(
        splashColor: Colors.black26,
        borderRadius: BorderRadius.circular(5),
        onTap: () {
          widget.onClick({widget.date: widget.events});
        },
        child: Card(
          margin: const EdgeInsets.all(0),
          shape: RoundedRectangleBorder(
            side: BorderSide(
                color: widget.isEnabled
                    ? widget.tileBorderColor
                    : Colors.grey.shade400,
                width: 0.5),
          ),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisSize: MainAxisSize.max,
            children: [
              SizedBox(
                child: Padding(
                  padding: const EdgeInsets.symmetric(
                      horizontal: 3.0, vertical: 2.0),
                  child: widget.todaysDate == widget.date
                      ? Container(
                          height: 24,
                          width: 24,
                          decoration: BoxDecoration(
                              borderRadius: BorderRadius.circular(100),
                              color: Colors.indigoAccent),
                          child: Align(
                            alignment: Alignment.center,
                            child: titleText(widget.title, 10,
                                textColor: widget.isEnabled
                                    ? Colors.white
                                    : Colors.grey,
                                textAlign: TextAlign.center),
                          ),
                        )
                      : titleText(widget.title, 10,
                          textColor: widget.isEnabled
                              ? widget.tileTextColor
                              : Colors.grey),
                ),
              ),
              Padding(
                  padding: const EdgeInsets.symmetric(horizontal: 2.0),
                  child: ListView.builder(
                      shrinkWrap: true,
                      itemCount:
                          widget.events.length >= 4 ? 4 : widget.events.length,
                      itemBuilder: (context, index) {
                        var item = widget.events[index];
                        return Card(
                          color: item.bgColor,
                          shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(3)),
                          elevation: 0,
                          margin: const EdgeInsets.only(bottom: 2),
                          child: Container(
                            alignment: Alignment.center,
                            width: double.infinity,
                            child: Padding(
                              padding: const EdgeInsets.symmetric(
                                  horizontal: 5.0, vertical: 2),
                              child: paragraphText(item.title,
                                  textColor: item.textColor, size: 7),
                            ),
                          ),
                        );
                      }))
            ],
          ),
        ),
      ),
    );
  }
}


lib/maxx_calendar.dart
library maxx_calendar;

import 'package:flutter/material.dart';
import 'package:flutter_animate/flutter_animate.dart';

import 'package:intl/intl.dart';
import 'package:maxx_calendar/widget/cal_model.dart';
import 'package:maxx_calendar/widget/calendar_dot_tile.dart';
import 'package:maxx_calendar/widget/calendar_tile.dart';
import 'package:maxx_calendar/widget/caltype.dart';
import 'package:maxx_calendar/widget/constants.dart';
import 'package:maxx_calendar/widget/views.dart';

//ignore: must_be_immutable
class MaxxCalendar extends StatefulWidget {
  DateTime now = DateTime.now();
  DateFormat formatter = DateFormat('EEE');
  DateFormat defaultFormat = DateFormat("dd-MM-yyyy");
  var calType = CalType.monthView;
  Color tileBorderColor = Colors.black54;
  Color tileDateColor = Colors.black54;
  var currMonth = DateTime.now().month;
  var currYear = DateTime.now().year;
  var firstDayOfMonth = "mon";
  var firstDaySpacing = 0;
  var endDaySpacing = 0;
  var emptyEvents = <List<MaxxEventModel>>[];
  var hideMonthNavigationButtons = false;

  var events = <List<MaxxEventModel>>[];

  var resultEvents = <List<MaxxEventModel>>[];
  Function(Map<String, List<MaxxEventModel>>) onDateClick;
  Function(String) onDateChangeListener;

  MaxxCalendar(
      {Key? key,
      required this.calType,
      required this.tileBorderColor,
      required this.tileDateColor,
      required this.events,
      required this.onDateClick,
      required this.onDateChangeListener,
      required this.currMonth,
      required this.currYear,
      required this.hideMonthNavigationButtons})
      : super(key: key);

  @override
  State<MaxxCalendar> createState() => _MaxxCalendarState();
}

class _MaxxCalendarState extends State<MaxxCalendar> {

  //get total no of days in the provided month
  static int getDaysInMonth(int year, int month) {
    if (month == DateTime.february) {
      final bool isLeapYear =
          (year % 4 == 0) && (year % 100 != 0) || (year % 400 == 0);
      return isLeapYear ? 29 : 28;
    }
    const List<int> daysInMonth = <int>[
      31,
      -1,
      31,
      30,
      31,
      30,
      31,
      31,
      30,
      31,
      30,
      31
    ];
    return daysInMonth[month - 1];
  }

  //creates the final events list for calendar
  void createFinalList() {
    widget.firstDayOfMonth =
        widget.formatter.format(DateTime(widget.currYear, widget.currMonth, 0));

    widget.firstDaySpacing = Constants.firstDayOffset(widget.currYear,
        (widget.currMonth), const DefaultMaterialLocalizations());

    if (widget.firstDaySpacing > 0) {
      for (var i = 1; i <= widget.firstDaySpacing; i++) {
        widget.emptyEvents.add([]);
      }
    }

    widget.resultEvents.addAll(widget.emptyEvents);

    if (widget.events.length <=
        getDaysInMonth(widget.currYear, widget.currMonth)) {
      widget.resultEvents.addAll(widget.events);
    } else {
      throw ("Events size : ${widget.events.length} - Events are more than the number of days present in current month");
    }

    var tempEndspacing = (getDaysInMonth(widget.currYear, widget.currMonth) +
        widget.firstDaySpacing);
    widget.endDaySpacing = 0;
    if ((tempEndspacing) % 7 != 0) {
      for (var i = 0; i < 10; i++) {
        widget.endDaySpacing++;
        if ((tempEndspacing + widget.endDaySpacing) % 7 == 0) {
          break;
        }
      }
    }
  }

  @override
  void initState() {
    super.initState();
    createFinalList();
  }

  @override
  Widget build(BuildContext context) {

    //handle next button click
    void onNextBtnClick() {
      setState(() {
        widget.resultEvents = [];
        widget.emptyEvents = [];
        if (widget.currMonth != 12) {
          widget.currMonth = widget.currMonth + 1;
        } else {
          widget.currMonth = 1;
          widget.currYear = widget.currYear + 1;
        }

        createFinalList();
      });
      widget.onDateChangeListener("${widget.currMonth}-${widget.currYear}");
    }

    //handle prev button click
    void onPrevBtnClick() {
      setState(() {
        widget.resultEvents = [];
        widget.emptyEvents = [];
        if(widget.currMonth!=1)
          {
            widget.currMonth = widget.currMonth - 1;
          }
        else
          {
            widget.currMonth = 12;
            widget.currYear = widget.currYear - 1;
          }

        createFinalList();
      });
      widget.onDateChangeListener("${widget.currMonth}-${widget.currYear}");
    }

    void invokeOnclick(Map<String, List<MaxxEventModel>> events) {
      widget.onDateClick(events);
    }

    return Column(
      children: [
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 5, vertical: 5),
          child: widget.hideMonthNavigationButtons == true
              ? const SizedBox()
              : Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    InkWell(
                      radius: 15,
                      borderRadius: BorderRadius.circular(15),
                      onTap: () {
                        onPrevBtnClick();
                      },
                      child: const Padding(
                        padding: EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.arrow_back_ios_new,
                          size: 15,
                        ),
                      ),
                    ),
                    paragraphText(
                        "${Constants.getCalendarMonth(widget.currMonth).toString()} ${widget.currYear}",
                        size: 12),
                    InkWell(
                      borderRadius: BorderRadius.circular(15),
                      child: const Padding(
                        padding: EdgeInsets.all(8.0),
                        child: Icon(
                          Icons.arrow_forward_ios,
                          size: 15,
                        ),
                      ),
                      onTap: () {
                        onNextBtnClick();
                      },
                    ),
                  ],
                ),
        ),
        SizedBox(
          height: 16,
          child: GridView.count(
            crossAxisCount: 7,
            children: [
              paragraphText("Mon", alignment: TextAlign.center),
              paragraphText("Tue", alignment: TextAlign.center),
              paragraphText("Wed", alignment: TextAlign.center),
              paragraphText("Thu", alignment: TextAlign.center),
              paragraphText("Fri", alignment: TextAlign.center),
              paragraphText("Sat", alignment: TextAlign.center),
              paragraphText("Sun", alignment: TextAlign.center),
            ],
          ),
        ),
        Expanded(
          flex: 9,
          child: Animate(
            effects: const [FadeEffect()],
            child: GridView.builder(
                gridDelegate: SliverGridDelegateWithFixedCrossAxisCount(
                    childAspectRatio:
                        widget.calType == CalType.monthView ? 8 / 16 : 1 / 1,
                    crossAxisSpacing: 0,
                    mainAxisSpacing: 0,
                    mainAxisExtent:
                        widget.calType == CalType.monthView ? 84 : 64,
                    crossAxisCount: 7),
                itemCount: getDaysInMonth(widget.currYear, widget.currMonth) +
                    widget.firstDaySpacing +
                    widget.endDaySpacing,
                itemBuilder: (BuildContext ctx, index) {
                  var item = ((widget.resultEvents.length) - 1) >= index
                      ? widget.resultEvents[index]
                      : <MaxxEventModel>[];
                  return Animate(
                    effects: [FadeEffect(delay: 500.ms)],
                    child: widget.calType == CalType.monthView
                        ? CalendarTile(
                            title: ((index - widget.firstDaySpacing < 0
                                    ? "${getDaysInMonth(widget.currMonth == 1 ? (widget.currYear - 1) : (widget.currYear), (widget.currMonth - 1 < 1 ? 12 : widget.currMonth - 1)) + 1 - (widget.firstDaySpacing) + index}"
                                    : index >=
                                            (getDaysInMonth(widget.currYear,
                                                    widget.currMonth) +
                                                widget.firstDaySpacing)
                                        ? "${index + 1 - (getDaysInMonth(widget.currYear, widget.currMonth) + widget.firstDaySpacing)}"
                                        : "${(index + 1) - widget.firstDaySpacing}"))
                                .toString(),
                            events: item,
                            onClick: (events) {
                              invokeOnclick(events);
                            },
                            isEnabled: (index - widget.firstDaySpacing < 0
                                ? false
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? false
                                    : true),
                            date: (index - widget.firstDaySpacing < 0
                                ? "" //previous month dates
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? "" //next month dates
                                    : "${index - widget.firstDaySpacing + 1 < 10 ? "0${index - widget.firstDaySpacing + 1}" : index - widget.firstDaySpacing + 1}-${widget.currMonth < 10 ? "0${widget.currMonth}" : widget.currMonth}-${widget.currYear}"),
                            todaysDate:
                                widget.defaultFormat.format(DateTime.now()),
                            tileBorderColor: widget.tileBorderColor,
                            tileTextColor: widget.tileDateColor,
                          )
                        : CalendarDotTile(
                            title: ((index - widget.firstDaySpacing < 0
                                    ? "${getDaysInMonth(widget.currMonth == 1 ? (widget.currYear - 1) : (widget.currYear), (widget.currMonth - 1 < 1 ? 12 : widget.currMonth - 1)) + 1 - (widget.firstDaySpacing) + index}"
                                    : index >=
                                            (getDaysInMonth(widget.currYear,
                                                    widget.currMonth) +
                                                widget.firstDaySpacing)
                                        ? "${index + 1 - (getDaysInMonth(widget.currYear, widget.currMonth) + widget.firstDaySpacing)}"
                                        : "${(index + 1) - widget.firstDaySpacing}"))
                                .toString(),
                            events: item,
                            onClick: (events) {
                              invokeOnclick(events);
                            },
                            isEnabled: (index - widget.firstDaySpacing < 0
                                ? false
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? false
                                    : true),
                            date: (index - widget.firstDaySpacing < 0
                                ? "" //previous month dates
                                : index >=
                                        (getDaysInMonth(widget.currYear,
                                                widget.currMonth) +
                                            widget.firstDaySpacing)
                                    ? "" //next month dates
                                    : "${index - widget.firstDaySpacing + 1 < 10 ? "0${index - widget.firstDaySpacing + 1}" : index - widget.firstDaySpacing + 1}-${widget.currMonth < 10 ? "0${widget.currMonth}" : widget.currMonth}-${widget.currYear}"),
                            todaysDate:
                                widget.defaultFormat.format(DateTime.now()),
                            tileBorderColor: widget.tileBorderColor,
                            tileTextColor: widget.tileDateColor,
                          ),
                  );
                }),
          ),
        ),
      ],
    );
  }
}
